### DISCUSSION TITLE
I can not match both two-letter and five-letter language codes from the `LANG` environment variable in an `if` statement in Nushell

### DISCUSSION THREAD

**Discussion by gusbemacbe** [nushell/nushell#13900] [Q&A] [answered]
## Explanation

It was easy to implement a translation system in Bash and Zsh. 

Now, it is particularly challenging to implement a translation system in Nushell because Nushell is different from Bash and Zsh. 

I am trying to match both two-letter language codes (like `en` for English, `pt` for Portuguese and European Portuguese) and five-letter language codes (like `pt_BR` for Brazilian Portuguese) correctly, and to fall back the non-existent languages to the English version. 

However, my current code only matches the two characters of the `LANG` environment value for all other languages.

Here is my small code:

```nu
let language = $env.LANG | str substring 0..1

let translations = {
  "pt":
  {
    "hello_world": "Olá, gajo",
  },
  "pt_BR":
  {
    "hello_world": "Oi, garoto",
  },
  _: 
  {
    "hello_world": "Hello, guy",
  }
}

let current_translations = $translations | get $lang_code
let hello_world = $current_translations.hello_world
```

I ran the following commands in the terminal, using Nushell:

```sh
$env.LANG = "pt_BR.UTF-8"; source 01-translations; print $hello_world
$env.LANG = "pt_PT.UTF-8"; source 01-translations; print $hello_world
$env.LANG = "es_ES.UTF-8"; source 01-translations; print $hello_world
```

I also tried to implement another alternative:

```nu
let language = $env.LANG

let lang_code = if $language == "pt_BR.UTF-8" {
  $language | str substring 0..5
} else {
  $language | str substring 0..1
}
```

## Expected behaviour

- The code correctly returns all two-letter language codes, including `pt` and displays the correct translation.
- For other two-letter languages codes that do not exist in the translation system, the code failed to fall back to the default English version. 
- For other five-letter languages codes, if you select `pt_BR.UTF-8`, it only returns the European Portuguese version incorrectly, causing the lookup in the translations table to fail.

## Desired behaviour

I want to handle both two-letter and five-letters language codes like `pt` and five-letter codes like `pt_BR`, and to fall back the non-existent two-letter and five-letters language codes to the default English version.

**Comment by fdncred** [comment]
Have you tried using `match`? Something like this.
```nushell
❯ $env.LANG
en_US.UTF-8
❯ match ($env.LANG | str substring 0..1) {
❯❯❯ "en" => { print "english" }
❯❯❯ _ => { print "other" }
❯❯❯ }
english
```

**Reply by gusbemacbe** [reply]
> ```nushell
> match ($env.LANG | str substring 0..1)
> ```

I tested the following code:

```nu
match ($env.LANG | str substring 0..1) {
  "pt" => { 
    let hello_world = "Olá"
  }
  "pt_BR" => { 
    let hello_world = "Oi"
  }
  _ => { 
    let hello_world = "Hello"
  }
}
```

This code failed to find the variable `hello_world`:

![image](https://github.com/user-attachments/assets/124ce75c-ae60-40f2-b0b7-2b448a850587)

I switched to `nu` instead of `source`, but it also failed:

![image](https://github.com/user-attachments/assets/f6daa27e-57bb-4048-90b7-b2de9ec76f8d)

**Reply by fdncred** [reply]
yup, that's right. you can't do it that way because the `let` variables are scoped. you'd have to do something more like this.
```nushell
> let output = match ($env.LANG | str substring 0..1) {
  "pt" => { 
    "Olá"
  }
  "pt_BR" => { 
    "Oi"
  }
  _ => { 
    "Hello"
  }
}
> print $output
```

**Reply by gusbemacbe** [reply]
The problem is that I want to create many variables by each language:

```
hello_world = "Hello"
bye_world = "Bye"
```

Here is the original Bash script code, that I tried to convert to Nushell:

```sh
curlang="${LANG:0:2}"

case "$curlang" in
pt) 
  hello_world="Olá"
  bye_world="Chau"
  ;;
pt_BR) 
  hello_world="Oi"
  bye_world="Tchau"
  ;;
*) 
  hello_world="Hello"
  bye_world="Bye"
  ;;
esac
```

**Reply by gusbemacbe** [reply]
I tested, and the code failed to detect five-letter language code like `pt_BR`:

![image](https://github.com/user-attachments/assets/846f8120-215d-4489-a93a-948f730c31a5)

**Reply by gusbemacbe** [reply]
I tested more another alternative, it worked, but the code is ugly:

```nu
let lang = if ($env.LANG =~ 'fr') { "fr" } else if ($env.LANG =~ 'pt_BR') { "pt_BR" } else if ($env.LANG =~ 'pt') { "pt" } else if ($env.LANG =~ 'it') { "it" } else { "default" }

let hello = $translations | get $lang | get hello
let bye = $translations | get $lang | get bye
let how_are_you = $translations | get $lang | get how_are_you
let where_are_you_from = $translations | get $lang | get where_are_you_from
```

**Reply by ghost** [reply]
> Here is the original Bash script code

i dont think that bash code works with both 5 and 2 either `"${LANG:0:2}"` removes `_BR` making the second case completely useless.

> I am trying to match both two-letter language codes

im not completely familiar with language-code specs, but its `AA_BB.encoding` with `_BB` and `.encoding` beeing optional, right? how about we use the `.` as seperator: `$env.LANG | default "EN" | split row '.' | get 0` (the `| default` is in case it is not set and would be required in your examples as well).

> I tested more another alternative

`let` is scoped to a block (`{ .. }`), but you can just let a block return something and use `let` above it (see fdncred's example).

you can also return dictionaries. and keeping all translations inside of 1 variable instead of filling up the variable-space is also a good practice in my opinion:

```nu
let T = (match ($env.LANG | default "EN" | split row '.' | get 0) {
  "DE" => {
    {
      "hello": "Hallo"
      "bye": "Tschüss"
    }
  }
  _ => {  # default (no need to define English twice -> put English in here)
    {
      "hello": "Hi"
      "bye": "Bye"
    }
  }
})

print $T.hello
```

another tip: use format strings (in some languages its not `hello name`, but `name hello`, or similar things)

```nu
let T = {"hello": "Hello {name}"}

{"name": "Bob"} | format pattern $T.hello
```

but also: i wouldn't recommend making cli-tools multi-language. its a lot of work (adding new languages, writing down whats called how, updating 30 things when changing one thing, etc), introduces a lot of bugs (wrong name used, etc), makes the program way less readable (over 50% of it is a translation-table), depending on the implementation makes it way slower, etc.

**Comment by NotTheDr01ds** [comment, marked as answer]
Yes, @Jan9103 is exactly right. I think the pattern you are looking for is to create a record with the translated phrases:

```nu
let translation = match $env.LANG {
  pt => {
    hello: 'Ol├í'
    bye: 'Chau'
  }

  pt_BR => {
    hello: 'Oi'
    bye: 'Tchau'
  }

  _ => {
    hello: 'Hello'
    bye: 'Goodbye'
  }
}
```

Then use `translation.<word>` such as:

```nu
$"I don't know why you say, "($translation.bye)", I say, "($translation.hello)" - The Beatles"
```
